### DISCUSSION TITLE
Raspberry Pi DSKY Build- Bit of help appreciated

### DISCUSSION THREAD

**Discussion by MKme** [virtualagc/virtualagc#1170] [General]
Hey all!  I'm Eric and over the last few months I've been building a 1:1 DSKY from another designer who posted the build on Hackaday https://hackaday.io/project/174524-accurate-apollo-dsky-replica   You can see my entire build in pics and videos (vids are still about 2 weeks behind) here: https://forum.mkme.org/viewtopic.php?f=4&t=1156

So I'm super happy with the build and the hardware is done but the software is kicking my butt.  After some trials I found a previous issue here which gave me enough to figure out how to build virtualagc from source on the Pi.  This is good and can run from the GUI but to get the Display/Arduinos talking is proving more difficult.  

The original code from Mano has 2 files (https://github.com/ManoDaSilva/AGC_DSKY_Replica/tree/master/Python%20Scripts) : 
1. runAGC.sh Bash script to fire things up
2. ManoDSKY.py to do all the fun stuff which seems to be based originally on piDSKY.py

The issues I have thwarting my beauty DSKY on each :) 
1. I cant get runAGC.sh to run in Raspbian as it has numerous errors even after I sudo chmod +x runAGC.sh 
- bash: ./runAGC.sh: /bin/bash^M: bad interpreter: No such file or directory
- I get more errors yet if I try altering the file by removing #!/bin/bash from the first liine so I gave up and proceeded to run things manually.  This is where I fail worse sadly..

2. After installing Python2 and pip install pyserial running: python2 manoDSKY.py 
Results is this:
manoDSKY.py:465: SyntaxWarning: name 'R1_1' is assigned to before global declaration global R1_1
manoDSKY.py:477: SyntaxWarning: name 'R1_1' is assigned to before global declaration global R1_1
manoDSKY.py:481: SyntaxWarning: name 'R1_1' is assigned to before global declaration global R1_1
manoDSKY.py:497: SyntaxWarning: name 'R2_1' is assigned to before global declaration global R2_1
manoDSKY.py:509: SyntaxWarning: name 'R2_1' is assigned to before global declaration global R2_1
manoDSKY.py:513: SyntaxWarning: name 'R2_1' is assigned to before global declaration global R2_1
manoDSKY.py:535: SyntaxWarning: name 'R3_1' is assigned to before global declaration global R3_1
manoDSKY.py:547: SyntaxWarning: name 'R3_1' is assigned to before global declaration global R3_1
manoDSKY.py:551: SyntaxWarning: name 'R3_1' is assigned to before global declaration global R3_1
Traceback (most recent call last):
  File "manoDSKY.py", line 54, in <module>
    import serial
ImportError: No module named serial

So I messaged Mano and he didnt recall having any trouble but he is away from his dev PC.  I had hoped someone here might have some thoughts I can try or even another way to interface virtualagc to the hardware I have running exactly as Mano does in his project.  The display is software serial on the Arduino Nano which also runs the annunciator LEDs.  The Arduino Pro Micro does the keyboard emulation to the Pi.  All should be working if I can just get the programs fired up as is done in the Python script

I tried running it with Python3 which results in the same first syntax error and I think the rest came one by one if I tried editing out the global declaration and putting it at the beginning (I could have done this wrong as Im pretty new to Python) 

So that's my novel.  Any ideas appreciated and thanks for all the work on this.  I cannot wait to share my functioning DSKY with the viewers on my YT channel.  So far many of them are just as excited as me with the project.  Thanks again for all the work.  

2 files in question attached
Eric 
[DSKY troublemaker scripts kicking Erics butt.zip](https://github.com/virtualagc/virtualagc/files/8393672/DSKY.troublemaker.scripts.kicking.Erics.butt.zip)

**Comment by rburkey2005** [comment]
Well, the errors you're indicating come from runAGC.sh look like they're characteristic of `bash` either not being installed at all, or else not being located at /bin/bash.  

What happens if you just try running

     bash

from the command line?  If that works okay, you should be able to do

     bash ./runAGC.sh

(rather than `./runAGC.sh`) even if the line "#!/bin/bash" is wrong.

If you try

     which bash

from the command line, it will at least tell you what folder it's in, and you can fix the line "#!/bin/bash".

**Reply by MKme** [reply]
Thanks Ron!  Truly appreciate you taking the time.  So running bash just returns nothing so seemed okay:

```
pi@dsky:~/Desktop $ bash ./runAGC.sh
Enabling numlock
setleds: Unrecognized argument: +num
Usage: setleds [option...] [[+|-][ num | caps | scroll ]]

Thus,
	setleds +caps -num
will set CapsLock, clear NumLock and leave ScrollLock unchanged.
The settings before and after the change (if any) are reported
when the -v option is given or when no change is requested.
Normally, setleds influences the vt flag settings
(and these are usually reflected in the leds).
With -L, setleds only sets the leds, and leaves the flags alone.
With -D, setleds sets both the flags and the default flags, so
that a subsequent reset will not change the flags.

Options:
  -D, --default   change both the VT flags and their default settings.
  -F, --flags     change the VT flags.
  -L, --leds      change only the leds.
  -V, --version   print version number.
  -v, --verbose   be more verbose.
  -h, --help      print this usage message.

Report bugs to authors.

Starting VirtualAGC
./runAGC.sh: line 5: screen: command not found
Starting DSKY
./runAGC.sh: line 7: cd: $'/home/pi/virtualagc/piPeripheral/\r': No such file or directory
': [Errno 2] No such file or directory
```

Screen command not found and Looks like some character in the file "r"??  I dont get it but running it this way does run.  Closer to winning now :)

**Comment by rburkey2005** [comment]
And I think that if `bash` turns out not to be installed, you can install it with something like

     apt install bash

**Reply by MKme** [reply]
It was indeed installed already and located (thanks again for looking!)

```
pi@dsky:~/Desktop $ which bash
/usr/bin/bash
```

**Comment by rburkey2005** [comment]
Text editing programs in Windows use a different way to delimit lines of text than Linux and Mac OS do.  Specifically, Linux and Mac OS, or in this case Raspbian, use a linefeed character by itself.  This is normally denoted as "\n".

Windows, on the other hand, delimits a line of text with carriage-return characters (normally denoted as "\r") followed by a linefeed ("\n").

So what the "\r" thing is telling you is that somehow a Windows carriage-return character got into runAGC.sh.

**Reply by MKme** [reply]
Fixed!  Bang on- had to make a new .sh file in editor- resaving it with any edits wouldnt scrub it

**Comment by rburkey2005** [comment]
As far as "screen" is concerned, do

     sudo apt install screen

**Reply by MKme** [reply]
Ugh I'm embarrassed by this.  I should have thought of this. BTW I saved all the install notes for Rasbian if you want the list for your tutorials  when Im done with my mess :)  

You nailed it- Down to only Python problems and "Inappropriate ioctl for device" now 

```
pi@dsky:~/Desktop $ bash ./test.sh
Enabling numlock
setleds: Unable to read keyboard flags: Inappropriate ioctl for device
Starting VirtualAGC
Starting DSKY
manoDSKY.py:465: SyntaxWarning: name 'R1_1' is assigned to before global declaration
  global R1_1
manoDSKY.py:477: SyntaxWarning: name 'R1_1' is assigned to before global declaration
  global R1_1
manoDSKY.py:481: SyntaxWarning: name 'R1_1' is assigned to before global declaration
  global R1_1
manoDSKY.py:497: SyntaxWarning: name 'R2_1' is assigned to before global declaration
  global R2_1
manoDSKY.py:509: SyntaxWarning: name 'R2_1' is assigned to before global declaration
  global R2_1
manoDSKY.py:513: SyntaxWarning: name 'R2_1' is assigned to before global declaration
  global R2_1
manoDSKY.py:535: SyntaxWarning: name 'R3_1' is assigned to before global declaration
  global R3_1
manoDSKY.py:547: SyntaxWarning: name 'R3_1' is assigned to before global declaration
  global R3_1
manoDSKY.py:551: SyntaxWarning: name 'R3_1' is assigned to before global declaration
  global R3_1
Traceback (most recent call last):
  File "manoDSKY.py", line 54, in <module>
    import serial
ImportError: No module named serial
```

**Comment by rburkey2005** [comment]
Well, the Python program is compatible with Python3, not with Python2, so that's one thing.  Python2 and Python3 are mutually incompatible in some ways, so you usually can't run a Python3 script with a Python2 interpreter, or vice-versa.

The `global` statements have to be present.  I'm surprised to see them located where they are (in the original code you linked), but admit I don't think there's anything _wrong_ with how they're placed.  Probably.  When you say you moved them to the "beginning", I'm not sure what you mean.  If you mean the beginning of the file, that's definitely wrong and won't work.  If you mean you moved them to the beginning of the `outputFromAGC` function, where you also see the line

     	global last10, last11, last13, last163, plusMinusState1, plusMinusState2, plusMinusState3

then that should have worked.  It's where I would normally put them.  If you move them, then that's where you should move them to.

**Reply by MKme** [reply]
You=Hero.  This was the key and I think the last major hurdle till I double check my hardware- I had incorrectly tried relocating to the beginning of the script (my limited C++ experience led me here) but with your suggestion I moved to outputFromAGC with the others.  This resulted in remaining errors but on removing them from the subsequent declarations- it seems to run!  I will check the hardware later on but it is indeed running the agc :)  

I am in your debt Sir

**Comment by rburkey2005** [comment]
Oh, and I don't know about the ioctl thing.  I don't know anything at all about `setleds`.  However, it's pretty obvious that it's only there to turn on NUM LOCK on the keyboard, so if you just press the NUM LOCK key, I doubt you need `setleds` for anything.  You might try running the command

     man setleds

for more info than what the error message gives you.  Or google the error message, being sure to include the keywords "setleds" and "raspbian" in your search query.

**Reply by MKme** [reply]
Found it in some posts on "enable numlock raspbian" :) Thanks again

**Comment by rburkey2005** [comment]
Actually, just running 

     setleds +num

on my own desktop computer (Linux, but not a raspberry pi), I get the error message

     KDGKBLED: Inappropriate ioctl for device
     Error reading current flags setting. Maybe you are not on the console?

so it's not an uncommon problem.

**Reply by MKme** [reply]
Got it!  setleds=+num is the key I think.  Stops the errors for now and seems legit

**Comment by rburkey2005** [comment]
I had failed to notice the error message "No module named serial".  You use the `pip3` program to install Python3 modules.  I think `pip3` is included with the Python3 installation and doesn't need to be installed separately.  Anyway, the `serial` module would presumably be installed with the command:

    pip3 install serial

I don't _think_ this needs `sudo`. 

For Python2, the corresponding installation program is just `pip`, but again, I'd encourage _not_ trying to use Python2 to run a Python3 script.

**Reply by MKme** [reply]
Yep I had installed it. I had tried to install for Python 2 but couldnt seem to get it- 3 is no problem and indeed I think you are right it needs to be run with 3 for sure.

**Comment by rburkey2005** [comment]
👍Sent on my Samsung Galaxy A51.
-------- Original message --------From: Eric ***@***.***> Date: 4/1/22  1:11 PM  (GMT-06:00) To: virtualagc/virtualagc ***@***.***> Cc: Ronald Burkey ***@***.***>, Comment ***@***.***> Subject: Re: [virtualagc/virtualagc] Raspberry Pi DSKY Build- Bit of help
  appreciated (Discussion #1170) 
You=Hero.  This was the key and I think the last major hurdle till I double check my hardware- I had incorrectly tried relocating to the beginning of the script (my limited C++ experience led me here) but with your suggestion I moved to outputFromAGC with the others.  This resulted in remaining errors but on removing them from the subsequent declarations- it seems to run!  I will check the hardware later on but it is indeed running the agc :)
I am in your debt Sir

**Comment by MKme** [comment]
Well I had it!  It worked and I debugged my bad USB for the keyboard emulation Arduino.  I had LED outputs but no software serial to the Nextion display so I swapped RX and TX then poof the Pi just derps out a refusal to connect to yagc.  I even swapped both to port 19698 from 19697 its using and nope. just 

```
pi@dsky:~/Desktop $ bash ./test.sh
Enabling numlock
Starting VirtualAGC
Starting DSKY
Keyboard echo off
declaring variables
LAMPS UPDATED
Could not connect to yaAGC (localhost:19697), exiting: [Errno 115] Operation now in progress
Could not connect to yaAGC (localhost:19697), exiting: [Errno 111] Connection refused
Could not connect to yaAGC (localhost:19697), exiting: [Errno 115] Operation now in progress
Could not connect to yaAGC (localhost:19697), exiting: [Errno 111] Connection refused
Could not connect to yaAGC (localhost:19697), exiting: [Errno 115] Operation now in progress
Exiting ...
```
So close!  It was alive.  I will win but maybe I have to rebuild the AGC sw.  Wonder if somehow it got corrupted when I was messing around with the peripherals.  Almost there.  Thanks so much for all the help before.  I will get it and win :)

**Comment by MKme** [comment]
Well make clean install fixed it.  Back to connecting fine and bonus- my display is working now with hardware RX and TX corrected.  

The kick in the pants now- my keyboard matrix is completely messed up.  All the numeric keys are somehow mapped wrong AND some not working at all so I have to dig in to that but its running :)

This is pretty incredible to me to have a piece of (simulated) Apollo tech running on my bench.  So happy and in your debt Ron.  Thanks for taking an interest and helping out.  All the Raspbian steps are pushed to the readme in my GIT fork and I'll have more video up next week and the week after as I work through the remainder.  IT RUNS!  :)
Cheers Sir

**Comment by rburkey2005** [comment]
Cool!

**Comment by rburkey2005** [comment]
Maybe it's the NUM LOCK messing you up.  :smile:

**Reply by MKme** [reply]
I actually think that it is :)  Looking in to it now.

**Reply by MKme** [reply]
It was the ASCII codes he was using in the Arduino ProMicro doing the keyboard input.  I have no idea why he had those numbers- they map to weird non-printable characters as I see it.  I just re-wrote it using normal letters and numbers and poof- The DSKY is completely alive in all respects.  A bit of latency in the serial com to work out but it starts and runs and accepts inputs.  The person running the keys has no clue so of course he gets nowhere but that is fiaxable with some reading and playing with your wonderful program now

Thanks again Ron.  Cant wait to play with this and learn more about the AGC/DSKY and HW.  Ive watched the CuriousMarc series 3 times now I think :)  

Cheers Sir.  I am in your debt
